src/internal/DotStarEsp32DmaSpiMethod.h

```diff
@@ -0,0 +1,320 @@
+    static const int parallelBits = 2;
+};
+
+class Esp32Hspi2BitBus
+{
+public:
+    static const spi_host_device_t spiHostDevice = HSPI_HOST;
+    static const int dmaChannel = 2;    // arbitrary assignment, but based on the fact there are only two DMA channels and two available SPI ports, we need to split them somehow
+    static const int parallelBits = 2;
+};
+
+class Esp32Vspi4BitBus
+{
+public:
+    static const spi_host_device_t spiHostDevice = VSPI_HOST;
+    static const int dmaChannel = 1;    // arbitrary assignment, but based on the fact there are only two DMA channels and two available SPI ports, we need to split them somehow
+    static const int parallelBits = 4;
+};
+
+class Esp32Hspi4BitBus
+{
+public:
+    static const spi_host_device_t spiHostDevice = HSPI_HOST;
+    static const int dmaChannel = 2;    // arbitrary assignment, but based on the fact there are only two DMA channels and two available SPI ports, we need to split them somehow
+    static const int parallelBits = 4;
+};
+
+template<typename T_SPISPEED, typename T_SPIBUS> class DotStarEsp32DmaSpiMethod
+{
+public:
+    typedef typename T_SPISPEED::SettingsObject SettingsObject;
+
+    DotStarEsp32DmaSpiMethod(uint16_t pixelCount, size_t elementSize, size_t settingsSize) :
+        _sizePixelData(pixelCount * elementSize + settingsSize),
+        _sizeEndFrame((pixelCount + 15) / 16) // 16 = div 2 (bit for every two pixels) div 8 (bits to bytes)
+    {
+        _spiBufferSize = _sizeStartFrame + _sizePixelData + _sizeEndFrame;
+
+        // must have a 4 byte aligned buffer for i2s
+        uint32_t alignment = _spiBufferSize % 4;
+        if (alignment)
+        {
+            _spiBufferSize += 4 - alignment;
+        }
+
+        _data = static_cast<uint8_t*>(malloc(_spiBufferSize));
+        _dmadata = static_cast<uint8_t*>(heap_caps_malloc(_spiBufferSize, MALLOC_CAP_DMA));
+
+        // data cleared later in NeoPixelBus::Begin()
+    }
+
+    // Support constructor specifying pins by ignoring pins
+    DotStarEsp32DmaSpiMethod(uint8_t, uint8_t, uint16_t pixelCount, size_t elementSize, size_t settingsSize) :
+        DotStarEsp32DmaSpiMethod(pixelCount, elementSize, settingsSize)
+    {
+    }
+
+    ~DotStarEsp32DmaSpiMethod()
+    {
+        if(_spiHandle) {
+            DeInitSpiDevice();
+            esp_err_t ret=spi_bus_free(T_SPIBUS::spiHostDevice);
+            ESP_ERROR_CHECK(ret);            
+        }
+        free(_data);
+        free(_dmadata);
+        _spiHandle = NULL;
+    }
+
+    bool IsReadyToUpdate() const
+    {
+        spi_transaction_t t;
+        spi_transaction_t * tptr = &t;
+        esp_err_t ret = spi_device_get_trans_result(_spiHandle, &tptr, 0);
+
+        // We know the previous transaction completed if we got ESP_OK, and we know there's no transactions queued if tptr is unmodified
+        return (ret==ESP_OK || tptr == &t);
```

Comment:
all operators have space before and after, so "ret==ESP_OK" shoud be "ret == ESP_OK"